### DISCUSSION TITLE
ROSCO Walkthrough Issues

### DISCUSSION THREAD

**Discussion by sautee** [NatLabRockies/ROSCO#152] [Q&A]
Hello,

I'm trying to run the ROSCO walkthrough locally - and I came across a few issues.

1. Tuning the controller requires optional params to be defined
2. Plotting PI gains throw a dimension exception error
3. When set up locally, loading the OpenFAST turbine model throws an error

**- Tune Controller Error**
```
# Tune controller 
controller      = ROSCO_controller.Controller(controller_params)
controller.tune_controller(turbine)
```

```
# Necessary parameters
---> 66         self.U_pc = list_check(controller_params['U_pc'], return_bool=False)
     67         self.zeta_pc = list_check(controller_params['zeta_pc'], return_bool=False)
     68         self.omega_pc = list_check(controller_params['omega_pc'], return_bool=False)

KeyError: 'U_pc'
```
This section threw a number of missing key errors. The NREL 5MW YAML file has optional control parameters not defined in it (like 'U_pc'). But it looks like the controller.py file does not have handlers set if the key isn't present in the YAML and throws a missing key error. Errors were thrown for -

- U_pc
- max_pitch
- ss_vsgain
- WS_GS_n and others

Once I defined all the params required in controller.py, this cell executed successfully.

**- Plotting PI gains throw a dimension exception error**. It looks like the numpy ndarrays have different dimensions.
```
# Plot gain schedule
fig, axs = plt.subplots(1, 2, sharey=True)
axs[0].plot(controller.v[len(controller.vs_gain_schedule.Kp):], controller.pc_gain_schedule.Kp)
axs[0].set_xlabel('Wind Speed')
axs[0].set_ylabel('Proportional Gain')
axs[0].grid('True')

axs[1].plot(controller.v[len(controller.vs_gain_schedule.Ki):], controller.pc_gain_schedule.Ki)
axs[1].set_xlabel('Wind Speed')
axs[1].set_ylabel('Integral Gain')
axs[1].grid('True')
```
```
ValueError                                Traceback (most recent call last)
[<ipython-input-112-8158ec24a967>](https://localhost:8080/#) in <module>()
      1 # Plot gain schedule
      2 fig, axs = plt.subplots(1, 2, sharey=True)
----> 3 axs[0].plot(controller.v[len(controller.vs_gain_schedule.Kp):], controller.pc_gain_schedule.Kp)
      4 axs[0].set_xlabel('Wind Speed')
      5 axs[0].set_ylabel('Proportional Gain')

[/usr/local/lib/python3.7/dist-packages/matplotlib/axes/_base.py](https://localhost:8080/#) in _plot_args(self, tup, kwargs)
    340 
    341         if x.shape[0] != y.shape[0]:
--> 342             raise ValueError(f"x and y must have same first dimension, but "
    343                              f"have shapes {x.shape} and {y.shape}")
    344         if x.ndim > 2 or y.ndim > 2:

ValueError: x and y must have same first dimension, but have shapes (31,) and (30,)

```

- **Error reading OpenFAST model**: I used a local Anaconda installation to download and run ROSCO. However, now I come across a new error. when trying to load the NREL 5MW OpenFAST model.

```
# Load turbine data from openfast model
turbine = ROSCO_turbine.Turbine(turbine_params)
turbine.load_from_fast(path_params['FAST_InputFile'],path_params['FAST_directory'],dev_branch=True,rot_source='txt',txt_filename=path_params['rotor_performance_filename'])
```

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-20-671a7bf61fbd> in <module>
      1 # Load turbine data from openfast model
      2 turbine = ROSCO_turbine.Turbine(turbine_params)
----> 3 turbine.load_from_fast(path_params['FAST_InputFile'],path_params['FAST_directory'],dev_branch=True,rot_source='txt',txt_filename=path_params['rotor_performance_filename'])

/srv/conda/envs/notebook/lib/python3.7/site-packages/ROSCO_toolbox/turbine.py in load_from_fast(self, FAST_InputFile, FAST_directory, FAST_ver, dev_branch, rot_source, txt_filename)
    180         fast.FAST_InputFile = FAST_InputFile
    181         fast.FAST_directory = FAST_directory
--> 182         fast.execute()
    183 
    184         if txt_filename:

/srv/conda/envs/notebook/lib/python3.7/site-packages/ROSCO_toolbox/ofTools/fast_io/FAST_reader.py in execute(self)
    396     def execute(self):
    397 
--> 398         self.read_MainInput()
    399         self.read_ElastoDyn()
    400         self.read_ElastoDynBlade()

/srv/conda/envs/notebook/lib/python3.7/site-packages/ROSCO_toolbox/ofTools/fast_io/FAST_reader.py in read_MainInput(self)
    482         self.fst_vt['Fst']['DT_Out'] = float_read(f.readline().split()[0])
    483         self.fst_vt['Fst']['TStart'] = float_read(f.readline().split()[0])
--> 484         self.fst_vt['Fst']['OutFileFmt'] = int(f.readline().split()[0])
    485         self.fst_vt['Fst']['TabDelim'] = bool_read(f.readline().split()[0])
    486         self.fst_vt['Fst']['OutFmt'] = f.readline().split()[0][1:-1]

ValueError: invalid literal for int() with base 10: '"NRELOffshrBsline5MW_Onshore_ServoDyn.dat"'
```
Help! 
Thanks.

**Comment by dzalkind** [comment]
Hi sautee,

Which branch/version of ROSCO are you using?  Please try on the latest `main` branch and I will do the same.

The inputs you point out are filled with defaults if not provided in the yaml, but this feature was only added in recent months. 

Best, Dan

**Reply by sautee** [reply]
Hi @dzalkind 

I'm using the package downloaded with Anaconda/conda-forge - which seems to be v2.4.0. I used the instructions under https://rosco.readthedocs.io/en/latest/source/install.html.

I just noticed that the latest version at conda-forge is 2.5.1. Somehow `conda install -y ROSCO` gets v2.4.

I'll try to build v2.5.1 locally or force conda to pull v2.5.1.

To note, I was able to resolve most issues I encountered earlier. I just have the dimension exception error when trying to plot PI gains. I'll install v2.5 and let you know if this persists.

Thanks for the help so far!

**Comment by sautee** [comment]
Hi @dzalkind 

I updated my local environment to ROSCO v2.5.1. 

- I can confirm that the 'optional' tuning parameters are not really optional. The controller.py has additional optional params in v2.5.1, and they throws new missing key errors. Moreover, 2.5.1 requires you to setup 'filter params' as well, which if not setup in the YAML file, throw errors as well.
```
File ~\anaconda3\envs\rosco-env\lib\site-packages\ROSCO_toolbox\controller.py:125, in Controller.__init__(self, controller_params)
    122     self.F_LPFDamping = 0.0
    124 # Filter parameters
--> 125 self.f_we_cornerfreq        = controller_params['filter_params']['f_we_cornerfreq']
    126 self.f_fl_highpassfreq      = controller_params['filter_params']['f_fl_highpassfreq']
    127 self.f_ss_cornerfreq        = controller_params['filter_params']['f_ss_cornerfreq']

KeyError: 'filter_params'
```
- The issue with array dimensions persists in 2.5.1. The wind array has one extra element vs the gain array.
```
File ~\anaconda3\envs\rosco-env\lib\site-packages\matplotlib\axes\_base.py:498, in _process_plot_var_args._plot_args(self, tup, kwargs, return_kwargs)
    495     self.axes.yaxis.update_units(y)
    497 if x.shape[0] != y.shape[0]:
--> 498     raise ValueError(f"x and y must have same first dimension, but "
    499                      f"have shapes {x.shape} and {y.shape}")
    500 if x.ndim > 2 or y.ndim > 2:
    501     raise ValueError(f"x and y can be no greater than 2D, but have "
    502                      f"shapes {x.shape} and {y.shape}")

ValueError: x and y must have same first dimension, but have shapes (31,) and (30,)
```

Hope this helps.

**Comment by dzalkind** [comment]
Hi,

I made some updates to the jupyter notebook here: #155 to handle this new default parameter feature.

Generally, I'd recommend using the example scripts because they are tested during each commit and pull request.  The examples cover everything in this notebook and more, and this notebook isn't part of the continuous integration, so we may take it out if we can't easily support it going forward. 

I hope this helps.  

Best, Dan

**Reply by sautee** [reply]
Thanks @dzalkind.

All previous errors are resolved. To note, I had to update my local wisdem package using `conda update wisdem`.

When running the 1-DOF simulation, I get the error -

`FileNotFoundError: Could not find module 'C:\Users\andre\rosco-test\ROSCO\build\libdiscon.dylib' (or one of its dependencies). Try using the full path with constructor syntax.`

Is there a build step I missed somewhere?

**Reply by sautee** [reply]
**Update** - I found the libdiscon.dll file under the `site-packages` folder of the conda environment. Once I pointed the notebook path appropriately, the 1DOF simulation ran successfully. Note - the default file name in the notebook `libdiscon.dylib` is for Mac environments.

I'm now at the OpenFAST simulation step. Here I get the error `AssertionError: File, ../Test_Cases/IEA-15-240-RWT-UMaineSemi/IEA-15-240-RWT-UMaineSemi.outb, does not exists`. 

How do I generate an `outb` file? (I couldn't find info in readthedocs)

**Reply by dzalkind** [reply]
You can generate an `outb` file using openfast.  An example can be found [here](https://github.com/NREL/ROSCO/blob/main/Examples/example_06.py).  You may need to edit some paths to work on your system, like the default dynamic library (.dll).  OpenFAST is good about error messages.

**Reply by sautee** [reply]
Thanks @dzalkind ! I'll give it a go and let you know how it goes.

**Comment by sautee** [comment]
When running example_06.py, I run into the following error.

```
Writing new controller parameter file parameter file: C:\Users\andre\rosco-test\Examples\DISCON.IN.
Using IEA-15-240-RWT-UMaineSemi.fst to run OpenFAST simulation
Running OpenFAST simulation for IEA-15-240-RWT-UMaineSemi.fst through the ROSCO toolbox...
Traceback (most recent call last):
  File "example_06.py", line 87, in <module>
    run_openfast(
  File "C:\Users\andre\anaconda3\envs\rosco-env\lib\site-packages\ROSCO_toolbox\utilities.py", line 525, in run_openfast
    subprocess.run([fastcall, os.path.join(fastfile)], check=True, cwd=cwd)
  File "C:\Users\andre\anaconda3\envs\rosco-env\lib\subprocess.py", line 493, in run
    with Popen(*popenargs, **kwargs) as process:
  File "C:\Users\andre\anaconda3\envs\rosco-env\lib\subprocess.py", line 858, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "C:\Users\andre\anaconda3\envs\rosco-env\lib\subprocess.py", line 1311, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
FileNotFoundError: [WinError 2] The system cannot find the file specified
```
It isn't clear to me which file it cannot find. Do I have to install OpenFAST separately to run this simulation? (I haven't installed OpenFAST yet).

**Reply by sautee** [reply]
FYI, I tried installing openfast in Windows, but it looks like the conda-forge openfast package is available only for MacOS and Linux environments.

**Reply by dzalkind** [reply]
I think you're right, you can only conda install openfast in mac/linux.

You can download executables from [here](https://github.com/OpenFAST/openfast/releases/tag/v3.2.0), and insert the path to the executable in the example_06.py code.

**Reply by sautee** [reply]
Thanks @dzalkind. I downloaded the OpenFAST v3.2.0 'openfast_x64.exe' executable from the link, and used it in example_06.py like so
```
fastcall = os.path.join(this_dir,'openfast_x64.exe')
run_openfast(
  os.path.join(this_dir,path_params['FAST_directory']),
  fastcall=fastcall, 
  fastfile=path_params['FAST_InputFile'], 
  chdir=True
  )
```
It start to run the simulation, but it fails looking for the **libdiscon.so** file. Since I'm running in Windows, shouldn't it be looking for **libdiscon.dll** file instead? Here's the full error

```
Running InflowWind.
Running ServoDyn.
Running ServoDyn Interface for Bladed Controllers (using Intel Visual Fortran for Windows).

FAST_InitializeAll:SrvD_Init:BladedInterface_Init:The dynamic library
.\../../ROSCO/install/lib/libdiscon.so could not be loaded. Check that the file exists in the
specified location and that it is compiled for 64-bit applications.

 OpenFAST encountered an error during module initialization.
 Simulation error level: FATAL ERROR

 Aborting OpenFAST.
```

**Reply by sautee** [reply]
The error with OpenFAST is resolved. I asked here https://github.com/OpenFAST/openfast/discussions/1179.

**You can change the path to the library in the ServoDyn input file. In your case, it is NRELOffshrBsline5MW_Onshore_ServoDyn.dat alongside your main OpenFAST input (.fst).**

**Comment by sautee** [comment]
Yet another Example_06.py question! (this is becoming a long thread!)

The assert in example_06 fails for me.

```
# Now double Kp_float and check that it's passed through
Kp_float = -18
controller_params['Kp_float'] = Kp_float
controller_params['tune_Fl'] = False
controller      = ROSCO_controller.Controller(controller_params)
controller.tune_controller(turbine)
np.testing.assert_almost_equal(Kp_float,controller.Kp_float)
```

I get this error. I understand what it is, but I can't figure out why the **Kp_float** value does not go through. I removed the assert test block to run the example.

This is the error.

```
Loading wind turbine data for NREL's ROSCO tuning and simulation processeses
-----------------------------------------------------------------------------
-----------------------------------------------------------------------------
   Tuning a reference wind turbine controller using NREL's ROSCO toolbox
-----------------------------------------------------------------------------
Loading FAST model: NREL-5MW.fst
Loading rotor performace data from text file: C:\Users\andre\rosco-test\Test_Cases\NREL-5MW\Cp_Ct_Cq.NREL5MW.txt
Loading rotor performace data from text file: C:\Users\andre\rosco-test\Examples\../Tune_Cases\../Test_Cases/NREL-5MW/Cp_Ct_Cq.NREL5MW.txt
-----------------------------------------------------------------------------
   Tuning a reference wind turbine controller using NREL's ROSCO toolbox
-----------------------------------------------------------------------------
Traceback (most recent call last):
  File "example_06.py", line 61, in <module>
    np.testing.assert_almost_equal(Kp_float,controller.Kp_float)
  File "C:\Users\andre\anaconda3\envs\rosco-env\lib\site-packages\numpy\testing\_private\utils.py", line 599, in assert_almost_equal
    raise AssertionError(_build_err_msg())
AssertionError:
Arrays are not almost equal to 7 decimals
 ACTUAL: -18
 DESIRED: 0.0
```

**Reply by dzalkind** [reply]
It looks like you've changed the example to run the NREL-5MW and it's not a floating model.  You can look at the code [here](https://github.com/NREL/ROSCO/blob/77278f5d1cee95a830b22d79797b351481e87972/ROSCO_toolbox/controller.py#L98) for more information.

**Reply by sautee** [reply]
Thanks @dzalkind, clear.

I ran OpenFAST for the NREL 5.0 MW Baseline Wind Turbine (Onshore) for 60 seconds. In the simulation out file, the X, Y and Z wind velocities remain constant through the simulation - X at 8 m/s, Y and Z at 0 m/s. 

Is there a way to generate changing wind speeds, or turbulent wind speeds in the simulation?

Here's what I get with example_08.py.
![08_NREL_5MW](https://user-images.githubusercontent.com/39371787/181073721-14a6fe54-6a96-4255-8194-c35e069d2b92.png)

**Reply by sautee** [reply]
Think I figured it out.

- The simulation time is specified in the .fst file with a default of 60 seconds. This value can be changed as necessary
- The 'InFlowWind' file (in my case 'NRELOffshrBsline5MW_InflowWind.dat') determines simulated wind conditions - steady state, uniform, turbulent etc. In in the InFlowWind file, I changed the WindType to 2 and pointed the config to a time series wind field file. Example InFlowWind files are available [here](https://openfast.readthedocs.io/en/dev/source/user/inflowwind/appendix.html) along with more detailed documentation [here](https://openfast.readthedocs.io/en/dev/source/user/inflowwind/index.html).

![08_NREL_5MW_2](https://user-images.githubusercontent.com/39371787/181088244-e6a91a65-5e0c-47e9-aed2-1fd1fe71c02c.png)

**Comment by mohsamaz** [comment]
Hello all,

I am using my own wind turbine and i can run openfast simulation.
also I can run default 02_ccblade.py example.

However, when i use 02_ccblade.py with my files, I get this error:

  File ~\ROSCO\rosco\toolbox\turbine.py:178 in load_from_fast
    fast.read_ElastoDynBlade(ed_blade_file)

  File ~\ROSCO\rosco\toolbox\ofTools\fast_io\FAST_reader.py:490 in read_ElastoDynBlade
    self.fst_vt['ElastoDynBlade']['BldFl2Sh'][i]  = float_read(f.readline().split()[0])

IndexError: list index out of range

could anyone please help regarding this issue?

Best regards,

**Reply by dzalkind** [reply]
This appears to be a mismatch between the OpenFAST input file version and the `FAST_reader`.  I'm not sure what version you are using for either, but it looks like you're using an older version of ROSCO.  Newer versions use the [`openfast-io` toolbox ](https://github.com/NREL/ROSCO/blob/03292150c047fe2904bf1316a76a6d26cee109ac/environment.yml#L25)for OpenFAST input/output processing.

I would check the format of your ElastoDynBlade file compared with the one read by default.

**Comment by mohsamaz** [comment]
Hello @dzalkind,
thank you for your reply.
I installed Rosco as an environment in anaconda and the version are:

rosco        2.9.6      pypi_0    pypi
openfast-io               4.1.2                    pypi_0    pypi
openfast-toolbox          3.5.1                     dev_0    <develop>

**Comment by mohsamaz** [comment]
Dear @dzalkind,

Apologies for following up again, but I’m still running into the same issue.

I’ve set up Rosco in an Anaconda environment with the following versions:

rosco 2.9.6 pypi_0 pypi

openfast-io 4.1.2 pypi_0 pypi

openfast-toolbox 3.5.1 dev_0

Unfortunately, the problem persists. Could you kindly provide some guidance or hints on how to resolve this?

Best regards,

**Reply by dzalkind** [reply]
I would ensure you're on the latest version of rosco (2.9.8).  Then start with a known, working example.

**Comment by mohsamaz** [comment]
Dear @dzalkind,
thank you for your prompt reply.

as I expain earlier I can run default 02_ccblade.py example without problem.

But with my openfast model I get error:

File ~\ROSCO\rosco\toolbox\turbine.py:178 in load_from_fast
fast.read_ElastoDynBlade(ed_blade_file)

File ~\ROSCO\rosco\toolbox\ofTools\fast_io\FAST_reader.py:490 in read_ElastoDynBlade
self.fst_vt['ElastoDynBlade']['BldFl2Sh'][i] = float_read(f.readline().split()[0])

IndexError: list index out of range

Although my OpenFAST model runs successfully(openfast simulation), I get this error when attempting to generate the aerodynamic coefficients of the turbine(by rosco). 

Could you advise on what might be causing this issue?

Best regards,

**Reply by dzalkind** [reply]
The latest version of ROSCO will use `openfast-io` as the `FAST_reader`, which should be consistent with your input file version.

As I said earlier, I would update to the the latest version of rosco (2.9.8), conda install openfast-io. Then start with a known, working example.

**Comment by mohsamaz** [comment]
Hello @dzalkind ,
thank you for latest update.
I updated :

# Name                    Version                   Build  Channel
rosco                          2.10.0                   pypi_0    pypi
rosco-toolbox             2.3.0                    pypi_0    pypi

But my previous problem persist. 
as I expain earlier I can run default 02_ccblade.py example without problem.

So, I made a new wind turbine model and i ran openfast  simulation without problem. now I want to use controller 
But with my own openfast model I get error:


  File ~\anaconda3\envs\rosco-env\lib\site-packages\openfast_io\FAST_reader.py:353 in read_MainInput
    self.fst_vt['Fst']['MHK'] = int(f.readline().split()[0])

ValueError: invalid literal for int() with base 10: '----------------------'

could you give me any hint to fix it,please?

Best regards,

**Reply by dzalkind** [reply]
Your OpenFAST input file should be consistent with this version of `openfast-io`: https://github.com/NREL/ROSCO/blob/159efac9be470110f4004aa3b17418936d5e1b84/environment.yml#L25
